crates/ruff_linter/src/rules/flake8_pyi/rules/generic_not_last_base_class.rs

```diff
@@ -0,0 +1,146 @@
+/// not the last base class in the bases list.
+///
+/// ## Why is this bad?
+/// `Generic[]` not being the final class in the bases tuple can cause
+/// unexpected behaviour at runtime (See [this CPython issue][1] for example).
+/// The rule is also applied to stub files, but, unlike at runtime,
+/// in stubs it is purely enforced for stylistic consistency.
+///
+/// For example:
+/// ```python
+/// class LinkedList(Generic[T], Sized):
+///     def push(self, item: T) -> None:
+///         self._items.append(item)
+///
+/// class MyMapping(
+///     Generic[K, V],
+///     Iterable[Tuple[K, V]],
+///     Container[Tuple[K, V]],
+/// ):
+///     ...
+/// ```
+///
+/// Use instead:
+/// ```python
+/// class LinkedList(Sized, Generic[T]):
+///     def push(self, item: T) -> None:
+///         self._items.append(item)
+///
+/// class MyMapping(
+///     Iterable[Tuple[K, V]],
+///     Container[Tuple[K, V]],
+///     Generic[K, V],
+/// ):
+///     ...
+/// ```
+/// ## References
+/// - [`typing.Generic` documentation](https://docs.python.org/3/library/typing.html#typing.Generic)
+///
+/// [1]: https://github.com/python/cpython/issues/106102
+#[violation]
+pub struct GenericNotLastBaseClass;
+
+impl Violation for GenericNotLastBaseClass {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("`Generic[]` should always be the last base class")
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Move `Generic[]` to the end".to_string())
+    }
+}
+
+/// PYI059
+pub(crate) fn generic_not_last_base_class(checker: &mut Checker, class_def: &StmtClassDef) {
+    let Some(bases) = class_def.arguments.as_deref() else {
+        return;
+    };
+
+    let semantic = checker.semantic();
+    if !semantic.seen_typing() {
+        return;
+    }
+
+    let generic_base_indices: Vec<usize> = bases
+        .args
+        .iter()
+        .enumerate()
+        .filter_map(|(base_index, base)| {
+            if semantic.match_typing_expr(map_subscript(base), "Generic") {
+                return Some(base_index);
+            }
+            None
+        })
+        .collect();
+
+    if generic_base_indices.is_empty() {
+        return;
+    }
+
+    let diagnostic = {
+        if generic_base_indices.len() == 1 {
+            let base_index = generic_base_indices[0];
+            if base_index == bases.args.len() - 1 {
+                // Don't raise issue for the last base.
+                return;
+            }
+
+            Diagnostic::new(GenericNotLastBaseClass, bases.range()).with_fix(generate_fix(
+                bases,
+                base_index,
+                checker.locator(),
+            ))
+        } else {
+            // No fix if multiple generics are seen in the class bases.
+            Diagnostic::new(GenericNotLastBaseClass, bases.range())
+        }
+    };
```

Comment:
There's quite a lot of `unwrap()` and `expect()` calls in the `generate_fix()` method currently. I think we could retrieve the node for the `Generic` base and the node for the last base in a more type-safe way, and get rid of some of those calls. That would mean that we'd be using the type system to our advantage rather than "fighting against it". You could maybe do something like this:

```rs
    let Some(last_base) = bases.args.last() else {
        return;
    };

    // Now we know both that the bases tuple is non-empty
    // and we've retrieved the node for the last base

    let mut generic_base_iter = bases
        .args
        .iter()
        .enumerate()
        .find(|(_, node)| semantic.match_typing_expr(map_subscript(node), "Generic"));

    let Some((generic_base_index, generic_base)) = generic_base_iter.next() else {
        return;
    }

    if generic_base_index == bases.args.len() - 1 {
        return;
    }

    let multiple_generic_bases = generic_base_iter.next().is_some();
```

You'd then have all the information you need to determine whether to emit a diagnostic or not (the index of the `Generic[]` base); all the information you need to determine whether it's fixable or not (whether or not there are multiple `Generic[]` bases -- though I'm still not _sure_ we should emit the lint at _all_ if there are); and you'd have extracted the two nodes you need for the autofix (the `Generic[]` base and the last base).

Reply:
The `.find()` should be `.filter()`, but impressively the rest of the code was pretty much spot on. Updated.